MekHQ/src/mekhq/gui/adapter/TOEMouseAdapter.java

```diff
@@ -109,42 +112,42 @@ public void actionPerformed(ActionEvent action) {
-                	if (singleForce.getAllUnits() !=null) {
-                	    String cantTech = "";
-                		for (UUID uuid : singleForce.getAllUnits()) {
-                			Unit u = gui.getCampaign().getUnit(uuid);
-                			if (u != null) {
-                			    if (null != u.getTech()) {
-                			        Person oldTech = u.getTech();
-                			        oldTech.removeTechUnitId(u.getId());
-                			        u.removeTech();
-                			    }
-                			    if (tech.canTech(u.getEntity())) {
-                			        u.setTech(tech.getId());
+            if (null != singleForce) {
+                Person tech = gui.getCampaign().getPerson(UUID.fromString(target));
+                if (null != tech) {
+                    if (singleForce.getTechID() != null) {
+                        Person oldTech = gui.getCampaign().getPerson(singleForce.getTechID());
+                        oldTech.clearTechUnitIDs();
+                        oldTech.addLogEntry(gui.getCampaign().getDate(), "Removed from " + singleForce.getName());
+                    }
+                    singleForce.setTechID(tech.getId());
+                    tech.addLogEntry(gui.getCampaign().getDate(), "Assigned to " + singleForce.getFullName());
+                    if (singleForce.getAllUnits() !=null) {
+                        String cantTech = "";
+                        for (UUID uuid : singleForce.getAllUnits()) {
+                            Unit u = gui.getCampaign().getUnit(uuid);
+                            if (u != null) {
+                                if (null != u.getTech()) {
+                                    Person oldTech = u.getTech();
+                                    oldTech.removeTechUnitId(u.getId());
+                                    u.removeTech();
+                                }
+                                if (tech.canTech(u.getEntity())) {
+                                    u.setTech(tech.getId());
                                     tech.addTechUnitID(u.getId());
                                     MekHQ.triggerEvent(new PersonTechAssignmentEvent(tech, u));
-                			    } else {
-                			        cantTech += tech.getName() + " cannot maintain " + u.getName() + "\n";
-                			    }
-                			}
-                		}
-                		if (cantTech != "") {
-                		    cantTech += "You will need to assign a tech manually.";
-                		    JOptionPane.showMessageDialog(null, cantTech, "Warning", JOptionPane.WARNING_MESSAGE);
-                		}
-                	}
-                }
-          	}
+                                } else {
+                                    cantTech += tech.getName() + " cannot maintain " + u.getName() + "\n";
+                                }
+                            }
+                        }
+                        if (cantTech != "") {
```

Comment:
.length() comparison instead?

Reply:
I didn't actually change this code, I just reformatted the spacing because it was wrong. I'm not going to change it either. Feel free to optimize it post-stable.